R/derive_param_wbc_abs.R

```diff
@@ -0,0 +1,172 @@
+#' "P02",    "LYMLE",   0.87,   "Lymphocytes (fraction of 1)",    "CYCLE 1 DAY 1",
+#' "P03",    "LYMLE",   0.89,   "Lymphocytes (fraction of 1)",    "CYCLE 1 DAY 1"
+#' )
+#'
+#' derive_param_wbc_abs (dataset = test_lb,
+#'                       by_vars = vars(USUBJID, VISIT),
+#'                       set_values_to = vars(
+#'                         PARAMCD = "LYMPH",
+#'                         PARAM = "Lymphocytes Abs (10^9/L)",
+#'                         DTYPE = "CALCULATION"
+#'                       ),
+#'                       get_unit_expr = extract_unit(PARAM),
+#'                       wbc_code = "WBC",
+#'                       diff_code = "LYMLE",
+#'                       diff_type = "fraction")
+
+derive_param_wbc_abs <- function(dataset,
+                                 by_vars,
+                                 set_values_to,
+                                 get_unit_expr,
+                                 wbc_unit = "10^9/L",
+                                 wbc_code = "WBC",
+                                 diff_code,
+                                 diff_type) {
+  assert_vars(by_vars)
+  assert_data_frame(dataset, required_vars = vars(!!!by_vars, PARAMCD, AVAL))
+  assert_character_scalar(wbc_code)
+  assert_character_scalar(diff_code)
+  assert_character_scalar(diff_type, values = c("percent", "fraction"))
+  get_unit_expr <- assert_expr(enquo(get_unit_expr))
+  assert_character_scalar(wbc_unit)
+  by_vars_str <- vars2chr(by_vars)
+
+  # Check unit of input parameters
+  assert_unit(dataset, wbc_code, required_unit = wbc_unit, get_unit_expr = !!get_unit_expr)
+
+  # Select observations and variables required for new observations.
+  dataset_temp <- dataset %>%
+    filter(
+      PARAMCD == !!wbc_code |
+      PARAMCD == !!diff_code |
+      PARAMCD == !!quo_get_expr(set_values_to$PARAMCD)
+    ) %>%
+    select(!!!by_vars, PARAMCD, AVAL)
+
+  # Only keep records where absolute value do not already exist.
+  dataset_abs <- dataset_temp %>%
+    filter(PARAMCD == !!quo_get_expr(set_values_to$PARAMCD)) %>%
+    mutate(temp_flag = "Y") %>%
+    select(!!!by_vars, temp_flag)
+
+  dataset_temp <- dataset_temp %>%
+    left_join(dataset_abs, by = by_vars_str) %>%
+    filter(is.na(temp_flag))
+
+  if (diff_type == "percent") {
+    analysis_value <- expr(!!sym(paste0("AVAL.", wbc_code)) *
+                             !!sym(paste0("AVAL.", diff_code)) / 100)
+  } else if (diff_type == "fraction") {
+    analysis_value <- expr(!!sym(paste0("AVAL.", wbc_code)) *
+                             !!sym(paste0("AVAL.", diff_code)))
+  }
+
+  # Create new parameter.
+  dataset_new <- dataset_temp %>%
+    derive_derived_param(
+      parameters = c(wbc_code,
+                     diff_code),
+      by_vars = by_vars,
+      analysis_value = !!analysis_value,
+      set_values_to = set_values_to
+    ) %>%
+    filter(PARAMCD == !!quo_get_expr(set_values_to$PARAMCD)) %>%
+    select(-starts_with("temp_"))
```

Comment:
Could we drop "temp_flag" when dataset_temp created above ie. immediately after its used. It may be overkill but what if user has a TEMP_xxx variable on their input data set? Maybe I am over-thinking.....

Reply:
Can probably ignore what I said about "TEMP_xxx" variables, just saw this in https://pharmaverse.github.io/admiral/articles/programming_strategy.html
"If the input dataset includes variables starting with temp_, an error must be issued."